### ISSUE TITLE
About BYOL implementations

### ISSUE THREAD

**Issue by wanggrun** [open-mmlab/mmselfsup#65]
https://github.com/open-mmlab/OpenSelfSup/blob/3272f765c5b7d5bee9772233a5a4d7e3fb66e5bf/openselfsup/models/byol.py#L92

Dear Xiaohang,

Thanks for the excellent job!

I have noticed that there are four feed-forwards in the following lines in BYOL.

    proj_online_v1 = self.online_net(img_v1)[0]
    proj_online_v2 = self.online_net(img_v2)[0]
    with torch.no_grad():
      proj_target_v1 = self.target_net(img_v1)[0].clone().detach()
      proj_target_v2 = self.target_net(img_v2)[0].clone().detach()

I also noticed that, in the earlier version, the pairs of images were concatenated such that only two feed-forwards were needed.
 
    img_cat1 = torch.cat([img_v1, img_v2], dim=0)
    img_cat2 = torch.cat([img_v2, img_v1], dim=0)
    proj_online = self.online_net(img_cat1)[0]
    with torch.no_grad():
      proj_target = self.target_net(img_cat2)[0].clone().detach()

My understanding is that these two implementations are similar except that the older one has a slight difference in calculating BN. However, the newer one is more time-consuming than the older one. Could you provide some explanations on the possible results of using the older implementation? As we know, you have remarked it as "fix BYOL" in the commit history.

I am looking forward to your reply.

**Comment by XiaohangZhan** [comment]
The way of computing BN is essential. The result of the old version is very poor. Maybe it is easy for the network to find some shortcut solution if they share BN.

**Comment by wanggrun** [comment]
Thank you for your reply!

I agree that BN may have found the shortcut solution. Years ago, after reading the following paper, I performed an experiment on CIFAR-10 and found that, with BN finding some shortcut solutions, the CIFAR-10 test set's error rate was almost reduced to zero.


Mohamed Hajaj, Duncan Gillies:
Batch Normalization and the impact of batch structure on the behavior of deep convolution networks. CoRR abs/1802.07590 (2018)
